homeassistant/components/dreo/coordinator.py

```diff
@@ -0,0 +1,150 @@
+"""Data update coordinator for Dreo devices."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any
+
+from hscloud.const import DEVICE_TYPE, FAN_DEVICE
+from hscloud.hscloud import HsCloud
+from hscloud.hscloudexception import HsCloudException
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util.percentage import ranged_value_to_percentage
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=10)
+
+
+class DreoGenericDeviceData:
+    """Base data for all Dreo devices."""
+
+    available: bool = False
+    is_on: bool = False
+
+    def __init__(self, available: bool = False, is_on: bool = False) -> None:
+        """Initialize generic device data."""
+        self.available = available
+        self.is_on = is_on
+
+
+class DreoFanDeviceData(DreoGenericDeviceData):
+    """Data specific to Dreo fan devices."""
+
+    mode: str | None = None
+    oscillate: bool | None = None
+    speed_percentage: int | None = None
+
+    def __init__(
+        self,
+        available: bool = False,
+        is_on: bool = False,
+        mode: str | None = None,
+        oscillate: bool | None = None,
+        speed_percentage: int | None = None,
+    ) -> None:
+        """Initialize fan device data."""
+        super().__init__(available, is_on)
+        self.mode = mode
+        self.oscillate = oscillate
+        self.speed_percentage = speed_percentage
+
+    @staticmethod
+    def process_fan_data(
+        device_model: str, status: dict[str, Any]
+    ) -> DreoFanDeviceData:
+        """Process fan device specific data."""
+
+        fan_data = DreoFanDeviceData(
+            available=status.get("connected", False),
+            is_on=status.get("power_switch", False),
+        )
+
+        if (mode := status.get("mode")) is not None:
+            fan_data.mode = str(mode)
+
+        if (oscillate := status.get("oscillate")) is not None:
+            fan_data.oscillate = bool(oscillate)
+
+        if (speed := status.get("speed")) is not None:
+            speed_range = (
+                FAN_DEVICE.get("config", {}).get(device_model, {}).get("speed_range")
+            )
+            if speed_range:
+                fan_data.speed_percentage = int(
+                    ranged_value_to_percentage(speed_range, float(speed))
+                )
+
+        return fan_data
+
+
+DreoDeviceData = DreoFanDeviceData | DreoGenericDeviceData
+
+
+class DeviceDataFactory:
+    """Factory to create device data based on device type."""
+
+    @staticmethod
+    def create_data(
```

Comment:
lets move the code in this method to `_async_update_data` or move this method into the Coordinator and drop this Factory class. There is no need for it now